File: src/swift/4_BehavioralPatterns/12_Visitor/Intrusive.swift
import Foundation

protocol Expression
{
  func print(_ buffer: inout String)
}

class DoubleExpression : Expression
{
  private var value: Double
  init(_ value: Double)
  {
    self.value = value
  }
  func print(_ buffer: inout String)
  {
    buffer.append(String(value))
  }
}

class AdditionExpression : Expression
{
  private var left: Expression
  private var right: Expression

  init(_ left: Expression, _ right: Expression)
  {
    self.left = left
    self.right = right
  }

  func print(_ buffer: inout String)
  {
    buffer.append("(")
    left.print(&buffer)
    buffer.append("+")
    right.print(&buffer)
    buffer.append(")")
  }
}

func main()
{
  let e = AdditionExpression(
    DoubleExpression(1),
    AdditionExpression(
      DoubleExpression(2),
      DoubleExpression(3)
    )
  )
  var s = ""
  e.print(&s)
  print(s)
}

main()

File: src/csharp/4_BehavioralPatterns/12_Visitor/Intrusive.cs
using System;
using System.Text;
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.Visitor.Intrusive
{
  public abstract class Expression
  {
    // adding a new operation
    public abstract void Print(StringBuilder sb);
  }

  public class DoubleExpression : Expression
  {
    private double value;

    public DoubleExpression(double value)
    {
      this.value = value;
    }

    public override void Print(StringBuilder sb)
    {
      sb.Append(value);
    }
  }

  public class AdditionExpression : Expression
  {
    private Expression left, right;

    public AdditionExpression(Expression left, Expression right)
    {
      this.left = left ?? throw new ArgumentNullException(paramName: nameof(left));
      this.right = right ?? throw new ArgumentNullException(paramName: nameof(right));
    }

    public override void Print(StringBuilder sb)
    {
      sb.Append(value: "(");
      left.Print(sb);
      sb.Append(value: "+");
      right.Print(sb);
      sb.Append(value: ")");
    }
  }

  public class Demo
  {
    private static void Main(string[] args)
    {
      var e = new AdditionExpression(
        left: new DoubleExpression(1),
        right: new AdditionExpression(
          left: new DoubleExpression(2),
          right: new DoubleExpression(3)));
      var sb = new StringBuilder();
      e.Print(sb);
      WriteLine(sb);

      // what is more likely: new type o rnew operation
    }
  }
}